File: Problems/No.841_Keys_and_Rooms/keys_and_rooms.py
class Solution(object):
    def canVisitAllRooms(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: bool
        """

        # parameters
        room_num, room_count = len(rooms), 0
        visited = [False] * room_num
        keys = []

        # start computing
        keys.append(0)
        while keys:
            key = keys.pop(0)
            if visited[key]:
                continue
            else:
                room_count += 1
                visited[key] = True
                for find_key in rooms[key]:
                    keys.append(find_key)

        return (room_count == room_num)
                    

File: Problems/No.841_Keys_and_Rooms/keys_and_rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        
        // parameters
        int room_num = rooms.size(), room_count = 0;
        std::vector<bool> visited(room_num, false);
        std::queue<int> keys;

        // start computing
        keys.push(0);
        while (!keys.empty()) {

            auto key = keys.front();
            keys.pop();
            
            if (visited[key] == true) continue;
            else {

                room_count++;
                visited[key] = true;
                for (auto find_key: rooms[key]) 
                    keys.push(find_key);

            }

        }

        return (room_count == room_num);

    }
};
